### ISSUE TITLE
Per-operation random seeds don't work in tensorflow 2.0

### ISSUE THREAD

**Issue by kazimuth** [tensorflow/tensorflow#33034]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Arch Linux
- TensorFlow installed from (source or binary): anaconda
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d38 2.0.0
- Python version: Python 3.7.4

**Describe the current behavior**

Minimal test script:

```python
import tensorflow as tf

x = tf.random.uniform([1], seed=1)
y = tf.random.uniform([1], seed=1)
print("x == y?", bool(tf.reduce_all(x == y)))
print(x)
print(y)
```

Output:
```
x == y? False
tf.Tensor([0.2390374], shape=(1,), dtype=float32)
tf.Tensor([0.22267115], shape=(1,), dtype=float32)
```

**Describe the expected behavior**

The script should print `x == y? True`.

According to the [docs](https://www.tensorflow.org/api_docs/python/tf/compat/v1/set_random_seed) on `tf.compat.v1.set_random_seed`:
>  If the graph-level seed is not set, but the operation seed is set: A default graph-level seed and the specified operation seed are used to determine the random sequence.

This does work correctly when not executing eagerly.

**Comment by gadagashwini** [comment]
@kazimuth,
I could replicate the issue with TF 2.0.0.rc2. Please see the colab [gist](https://colab.sandbox.google.com/gist/gadagashwini/847114c631edb377542a001d0c6714e5/untitled174.ipynb). Thanks!

**Comment by rmothukuru** [comment]
@kazimuth,
Thank you for the details and the reference. I think the behavior is expected, as per [the point](https://www.tensorflow.org/api_docs/python/tf/compat/v1/set_random_seed), 

> If the graph-level seed is not set, but the operation seed is set: A default graph-level seed and the specified operation seed are used to determine the random sequence. 

It is because, in Eager Execution, there will not be Graph Level Seed, as there will be no graph at all. So, they will be two independent Random Tensor Operations.

Please let me know your opinion about the same. Thanks.

**Comment by kazimuth** [comment]
I mean, that makes sense, but it's not really intuitive; it seems like the implementation details are getting in the way of the expected interface. Ideally it would be possible to set global and per-operation seeds in eager mode, whether or not there's a graph in the background.

In particular, as far as I can tell I can't write code with reproducible randomness right now if I'm running in eager mode; that makes reproducing experiments and writing tests harder.

**Comment by wangpengmit** [comment]
The behavior is caused by the facts that eager runtime caches and reuses a kernel for an op if the op is called multiple times with the same arguments, and that `tf.random.uniform` is a stateful op. In the line `x = ...`, a `tf.random.uniform` kernel is used and its internal state changed; in the line `y = ...`, the same kernel is used again but its internal state has been changed. The two `tf.random.uniform` will only produce identical numbers in eager mode if you put `tf.random.set_seed(...)` before each of them. The updated docstring for `set_seed` (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/random_seed.py#L191) explains those caveats. Alternatively you can use stateless random ops (https://www.tensorflow.org/api_docs/python/tf/random/stateless_uniform) or the new stateful random ops (https://www.tensorflow.org/api_docs/python/tf/random/experimental/Generator).

**Comment by kazimuth** [comment]
Ohh, I see, so this is intentional behavior. If I rerun my test script, I get the same tensors every run, good. Okay, in that case, this isn't a bug. Thanks for the clarification!

**Comment by monk1337** [comment]
In Tensorflow 2.0 you can set random seed like this :


    import tensorflow as tf
    tf.random.set_seed(221)


    from tensorflow import keras
    from tensorflow.keras import layers
    
    
    model = keras.Sequential( [ 
    layers.Dense(2,name = 'one'),
    layers.Dense(3,activation = 'sigmoid', name = 'two'),
    layers.Dense(2,name = 'three')])
    
    x = tf.random.uniform((12,12))
    model(x)

**Comment by irmaR** [comment]
> In Tensorflow 2.0 you can set random seed like this :
> 
> ```
> import tensorflow as tf
> tf.random.set_seed(221)
> 
> 
> from tensorflow import keras
> from tensorflow.keras import layers
> 
> 
> model = keras.Sequential( [ 
> layers.Dense(2,name = 'one'),
> layers.Dense(3,activation = 'sigmoid', name = 'two'),
> layers.Dense(2,name = 'three')])
> 
> x = tf.random.uniform((12,12))
> model(x)
> ```

Are you sure  "tf.random.set_seed(221)" works in Tensorflow 2.0? One of my systems has Tensorflow 2.0 and the deep learning there pipeline always returns a different result. The same code in Tensorflow 2.2 always returns the same result.
